core/iwasm/common/gc/gc_common.c

```diff
@@ -165,31 +214,107 @@ wasm_struct_obj_new_with_type(WASMExecEnv *exec_env, WASMStructType *type)
 }
 
 WASMArrayObjectRef
-wasm_array_obj_new_with_typeidx(WASMExecEnv *exec_env, uint32 type_idx)
-{
-    /* TODO */
+wasm_array_obj_new_with_typeidx(WASMExecEnv *exec_env, uint32 type_idx,
+                                uint32 length, wasm_value_t *init_value)
+{
+    WASMArrayObjectRef array_obj;
+    WASMRttTypeRef rtt_type;
+    WASMModuleCommon *module = wasm_exec_env_get_module(exec_env);
+    WASMModuleInstanceCommon *module_inst =
+        wasm_exec_env_get_module_inst(exec_env);
+    WASMModule *wasm_module = (WASMModule *)module;
+    WASMType *defined_type = wasm_get_defined_type(module, type_idx);
+
+    if (!(rtt_type = wasm_rtt_type_new(
+              defined_type, type_idx, wasm_module->rtt_types,
+              wasm_module->type_count, &wasm_module->rtt_type_lock))) {
+        wasm_set_exception((WASMModuleInstance *)module_inst,
+                           "create rtt type failed");
+        goto got_exception;
+    }
+
+    array_obj = wasm_array_obj_new(exec_env, rtt_type, length, init_value);
+
+    return array_obj;
+
+got_exception:
     return NULL;
 }
 
 WASMArrayObjectRef
-wasm_array_obj_new_with_type(WASMExecEnv *exec_env, WASMArrayType *type)
-{
-    /* TODO */
-    return NULL;
+wasm_array_obj_new_with_type(WASMExecEnv *exec_env, WASMArrayType *type,
+                             uint32 length, wasm_value_t *init_value)
+{
+    WASMArrayObjectRef array_obj;
+    uint32 type_idx = 0;
+    WASMModule *wasm_module = (WASMModule *)wasm_exec_env_get_module(exec_env);
+
+    for (uint32 i = 0; i < wasm_module->type_count; i++) {
+        if (wasm_module->types[i]->type_flag != WASM_TYPE_ARRAY) {
+            continue;
+        }
```

Comment:
Need to add the check, add check below and bh_assert after the for loop should be enough:
```C
    for (i = 0; i < wasm_module->type_count; i++) {
        if (wasm_module->types[i] == (WASMType *)type) {
            break;
        }
    }
    bh_assert(i < wasm_module->type_count);
    bh_assert(type->type_flag == WASM_TYPE_ARRAY);

    array_obj =
        wasm_array_obj_new_with_typeidx(exec_env, i, length, init_value);

    return array_obj;
```

Reply:
Same as above, should handle different case of module: module can be WASMModule of interpreter, or AOTModule of AOT